### PR TITLE
[FrontEnd]enable PDPD 16 ops conversion Part2

### PR DESCRIPTION
**Pull request by zhangYiIntel**
### Details:
 - *Enable 16 Ops Conversion*
| [bilinear\_interp](https://github.com/PaddlePaddle/Paddle/blob/758dd7bb8f560e45bc49855e4776a9d64c95aea9/paddle/fluid/operators/interpolate_v2_op.cc#L458)                  |
| [bmm](https://github.com/PaddlePaddle/Paddle/blob/758dd7bb8f560e45bc49855e4776a9d64c95aea9/paddle/fluid/operators/bmm_op.cc#L80)                               |
| [conv2d\_transpose](https://github.com/PaddlePaddle/Paddle/blob/758dd7bb8f560e45bc49855e4776a9d64c95aea9/paddle/fluid/operators/conv_transpose_op.cc#L245)                 |
| [dropout](https://github.com/PaddlePaddle/Paddle/blob/758dd7bb8f560e45bc49855e4776a9d64c95aea9/paddle/fluid/operators/dropout_op.cc#L49)                           |
| [equal](https://github.com/PaddlePaddle/Paddle/blob/69d8027471f82d9f73234a7df9a9396dec982ad5/paddle/fluid/operators/controlflow/compare_op.h#L50)                             |
| [expand\_v2](https://github.com/PaddlePaddle/Paddle/blob/758dd7bb8f560e45bc49855e4776a9d64c95aea9/paddle/fluid/operators/expand_v2_op.cc#L111)                        |
| [fill\_constant](https://github.com/PaddlePaddle/Paddle/blob/758dd7bb8f560e45bc49855e4776a9d64c95aea9/paddle/fluid/operators/fill_constant_op.cc#L85)                    |
| [fill\_constant\_batch\_size\_like](https://github.com/PaddlePaddle/Paddle/blob/758dd7bb8f560e45bc49855e4776a9d64c95aea9/paddle/fluid/operators/fill_constant_batch_size_like_op.cc#L32) |
| [flatten\_contiguous\_range](https://github.com/PaddlePaddle/Paddle/blob/69d8027471f82d9f73234a7df9a9396dec982ad5/paddle/fluid/operators/flatten_op.cc#L312)        |
| [greater\_equal](https://github.com/PaddlePaddle/Paddle/blob/69d8027471f82d9f73234a7df9a9396dec982ad5/paddle/fluid/operators/controlflow/compare_op.h#L44)                    |
| [hard\_swish](https://github.com/PaddlePaddle/Paddle/blob/69d8027471f82d9f73234a7df9a9396dec982ad5/paddle/fluid/operators/activation_op.h#L1000)                       |
| [nearest\_interp](https://github.com/PaddlePaddle/Paddle/blob/758dd7bb8f560e45bc49855e4776a9d64c95aea9/paddle/fluid/operators/interpolate_v2_op.cc#L458)                   |
| [leakyrelu](https://github.com/PaddlePaddle/Paddle/blob/69d8027471f82d9f73234a7df9a9396dec982ad5/paddle/fluid/operators/activation_op.h#L1144)                         |
| [log](https://github.com/PaddlePaddle/Paddle/blob/69d8027471f82d9f73234a7df9a9396dec982ad5/paddle/fluid/operators/activation_op.h#L828)                               |
| [logical\_not](https://github.com/PaddlePaddle/Paddle/blob/69d8027471f82d9f73234a7df9a9396dec982ad5/paddle/fluid/operators/controlflow/logical_op.h#L55)                      |
| [lstm](https://github.com/PaddlePaddle/Paddle/blob/758dd7bb8f560e45bc49855e4776a9d64c95aea9/paddle/fluid/operators/lstm_op.cc#L124)                              |
| [matmul](https://github.com/PaddlePaddle/Paddle/blob/758dd7bb8f560e45bc49855e4776a9d64c95aea9/paddle/fluid/operators/matmul_op.cc#L694)                            |
| [transpose2](https://github.com/PaddlePaddle/Paddle/blob/758dd7bb8f560e45bc49855e4776a9d64c95aea9/paddle/fluid/operators/matmul_op.cc#L694)                        |
### Tickets:
 - *53052*

### CODE REVIEW

File: ngraph/frontend/paddlepaddle/src/op_table.cpp [hunk 1]
```diff
@@ -8,10 +8,27 @@
...
+#include "op/fill_constant_batch_size_like.hpp"
+#include "op/flatten_contiguous_range.hpp"
+#include "op/greater_equal.hpp"
+#include "op/hard_sigmoid.hpp"
+#include "op/hard_swish.hpp"
+#include "op/interp.hpp"
+#include "op/leakyrelu.hpp"
+#include "op/log.hpp"
+#include "op/logical_not.hpp"
+#include "op/matmul.hpp"
```
Comment (ilya-lavrenov) [inline ngraph/frontend/paddlepaddle/src/op_table.cpp:26]: Can we use a forward declaration for the functions instead of header file with a single function?

Comment (zhangYiIntel) [reply ngraph/frontend/paddlepaddle/src/op_table.cpp:26]: @nosovmik What do you think ?

**Comment by ilyachur** [comment]
@jane-intel @mvafin Could you have a look?

File: ngraph/frontend/paddlepaddle/src/op/hard_sigmoid.cpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        {
+            namespace op
+            {
+                NamedOutputs hard_sigmoid(const NodeContext& node)
+                {
+                    auto data = node.get_ng_input("X");
+                    auto dtype = data.get_element_type();
+                    float slope = node.get_attribute<float>("slope", 0.2f);
+                    float offset = node.get_attribute<float>("offset", 0.5f);
+                    auto alpha = builder::make_constant(dtype, {}, slope);
```
Comment (mvafin) [inline ngraph/frontend/paddlepaddle/src/op/hard_sigmoid.cpp:24]: please don't use builder API, it creates broadcast nodes and you don't need it in this case.

**Review by mvafin — APPROVED** [review]

File: ngraph/frontend/paddlepaddle/src/op_table.cpp (hunk 1 above)
Comment (ilya-lavrenov) [reply ngraph/frontend/paddlepaddle/src/op_table.cpp:26]: @nosovmik ping.

We have a lot of dummy .hpp files with declarations only..

Comment (nosovmik) [reply ngraph/frontend/paddlepaddle/src/op_table.cpp:26]: @ilya-lavrenov Do you suggest to move all of them to single file, like op_table.hpp?
It would be easier to read, however each operation, like log.cpp, will need to include "op_table.hpp" which will have a lot of declarations not needed for log.cpp.
I'm basically ok with both approaches

Comment (ilya-lavrenov) [reply ngraph/frontend/paddlepaddle/src/op_table.cpp:26]: I suggest in current file:
```c++
namespace ngraph
{
    namespace frontend
    {
        namespace pdpd
        {
            namespace op
            {
                NamedOutputs greater_equal(const NodeContext& node);
                NamedOutputs transpose2(const NodeContext& node);
                // ..
            }
        } // namespace pdpd
    }     // namespace frontend
} // namespace ngraph 
```
So, we don't need any hpp files with single declaration

Comment (ilya-lavrenov) [reply ngraph/frontend/paddlepaddle/src/op_table.cpp:26]: Or we can use an approach like https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/src/cldnn_engine/ops/scatter_elements_update.cpp#L66

here we don't need a lot of .hpp files as well and each operation registers itself

**Comment by nosovmik** [comment]
Any reason of not including test models here? Just wanted to check if all cases are covered by tests

**Comment by zhangYiIntel** [comment]
Hi @nosovmik The tests cases are removed here because previously, we have different opinions about how to test conversion. Therefore we remove them in this PR and test them locally. I think we can add them back in another PR because we have alignment with Ilya about how to test the conversion.

File: ngraph/frontend/paddlepaddle/src/op_table.cpp (hunk 1 above)
Comment (nosovmik) [reply ngraph/frontend/paddlepaddle/src/op_table.cpp:26]: Yes, I guess something like dynamic ops registration may be needed in future if we'll allow to dynamically register some custom user-defined operations, but I guess not now :-)

File: ngraph/frontend/paddlepaddle/src/op/equal.cpp [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        namespace pdpd
+        {
+            namespace op
+            {
+                NamedOutputs equal(const NodeContext& node)
+                {
+                    auto data_x = node.get_ng_input("X");
+                    auto data_y = node.get_ng_input("Y");
+                    return node.default_single_output_mapping(
+                        {std::make_shared<ngraph::opset6::Equal>(data_x, data_y)}, {"Out"});
```
Comment (jane-intel) [inline ngraph/frontend/paddlepaddle/src/op/equal.cpp:21]: What about broadcasting? Equal has numpy broadcasting by default.

File: ngraph/frontend/paddlepaddle/src/op/expand_v2.cpp [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+namespace ngraph
+{
+    namespace frontend
+    {
+        namespace pdpd
+        {
+            namespace op
+            {
+                NamedOutputs expand_v2(const NodeContext& node)
```
Comment (jane-intel) [inline ngraph/frontend/paddlepaddle/src/op/expand_v2.cpp:17]: From the doc string [here](https://github.com/PaddlePaddle/Paddle/blob/758dd7bb8f560e45bc49855e4776a9d64c95aea9/paddle/fluid/operators/expand_v2_op.cc#L138). Expand could transform tensor of shape  [2, 3, 1] to the tensor of shape [2, 6, 1]. Dimension 3 is not broadcast-able to 6, please see the [broadcast_rules](https://github.com/openvinotoolkit/openvino/blob/master/docs/ops/broadcast_rules.md). According to what I have seen in the doc -- the operation could be expressed through openvino Tile operation -- 
https://github.com/openvinotoolkit/openvino/blob/master/docs/ops/movement/Tile_1.md

File: ngraph/frontend/paddlepaddle/src/op/flatten_contiguous_range.cpp [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+                            std::make_shared<opset6::StridedSlice>(shape_of_x,
+                                                                   axis2_begin,
+                                                                   axis2_end,
+                                                                   std::vector<int64_t>{0},
+                                                                   std::vector<int64_t>{0});
+                        axes.push_back(axis2_node);
+                    }
+
+                    auto new_shape_node = std::make_shared<opset6::Concat>(axes, 0);
+                    return node.default_single_output_mapping(
```
Comment (jane-intel) [inline ngraph/frontend/paddlepaddle/src/op/flatten_contiguous_range.cpp:50]: By setting special_zero to true you make Reshape to copy input dimension placed under the dimension where 0 is placed in the reshape_pattern input. Is this really what you wanted?

File: ngraph/frontend/paddlepaddle/src/op/hard_swish.cpp [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        {
+            namespace op
+            {
+                NamedOutputs hard_swish(const NodeContext& node)
+                {
+                    auto data = node.get_ng_input("X");
+                    if (node.has_attribute<float>("threshold"))
+                    {
+                        auto threshold = node.get_attribute<float>("threshold");
+                        PDPD_ASSERT(std::abs(threshold - 6.0) < 0.001,
```
Comment (nosovmik) [inline ngraph/frontend/paddlepaddle/src/op/hard_swish.cpp:23]: I have some doubts about this threshold. Consider model like
```
x1  = hard_swish(threshold=6.0, scale=3.0, offset=3.0)
x2  = hard_swish(threshold=6.0001, scale=3.0001, offset=3.0001)
out = (x2 - x1) * 10000
```
Execution of this will produce significant non-null values. But in OV it will be always 0

Comment (zhangYiIntel) [reply ngraph/frontend/paddlepaddle/src/op/hard_swish.cpp:23]: The OV hard_swish only supports fixed value of threshold(6.0) and offset(3.0). The check here intends to ensure that the Paddle Ops have those default value. Moreover, we don't plan to support other value because the Paddle has also fixed their threshold to 6 and offset to 3 in [API 2.1](https://www.paddlepaddle.org.cn/documentation/docs/en/api/paddle/nn/functional/hardswish_en.html).

File: ngraph/frontend/paddlepaddle/src/op/rnn.cpp [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    namespace frontend
+    {
+        namespace pdpd
+        {
+            namespace op
+            {
+                NamedOutputs rnn(const NodeContext& node)
+                {
+                    auto mode = node.get_attribute<std::string>("mode");
+                    PDPD_ASSERT(mode == "LSTM", "RNN only support LSTM now");
```
[the remark this answers is not in the document]
Comment (zhangYiIntel) [reply ngraph/frontend/paddlepaddle/src/op/rnn.cpp:21]: If user uses ops other than LSTM, the PDPD_ASSERT will fail and the program will print the `RNN only support LSTM now` and exit.

File: ngraph/frontend/paddlepaddle/src/op/transpose2.cpp [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+                NamedOutputs transpose2(const NodeContext& node)
+                {
+                    auto data = node.get_ng_input("X");
+                    auto perm = node.get_attribute<std::vector<int>>("axis");
+
+                    auto rank =
+                        static_cast<unsigned long>(data.get_partial_shape().rank().get_length());
+
+                    std::cout << perm.size() << std::endl;
+                    std::cout << data.get_partial_shape().rank() << ":" << rank << std::endl;
```
[the remark this answers is not in the document]
Comment (zhangYiIntel) [reply ngraph/frontend/paddlepaddle/src/op/transpose2.cpp:25]: will remove those useless code

File: ngraph/frontend/paddlepaddle/src/op/expand_v2.cpp (hunk 4 above)
Comment (zhangYiIntel) [reply ngraph/frontend/paddlepaddle/src/op/expand_v2.cpp:17]: @luo-cheng2021  Please have a fix

File: ngraph/frontend/paddlepaddle/src/op/dropout.cpp [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+        namespace pdpd
+        {
+            namespace op
+            {
+                NamedOutputs dropout(const NodeContext& node)
+                {
+                    auto data = node.get_ng_input("X");
+                    auto dropout_implementation =
+                        node.get_attribute<std::string>("dropout_implementation");
+                    if (dropout_implementation == "downgrade_in_infer")
```
[the remark this answers is not in the document]
Comment (zhangYiIntel) [reply ngraph/frontend/paddlepaddle/src/op/dropout.cpp:21]: @mangguo321 Please have a fix

File: ngraph/frontend/paddlepaddle/src/op/rnn.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (jane-intel) [reply ngraph/frontend/paddlepaddle/src/op/rnn.cpp:21]: What will it indicate to the user? As I see there is nothing connected to Paddle Paddle in the message. Can we narrow down the number of problems user would consider if he/she sees it? We have various RNN layers in the openvino and user could give up on us at this point

[the remark this answers is not in the document]
Comment (zhangYiIntel) [reply ngraph/frontend/paddlepaddle/src/op/rnn.cpp:21]: Yep, you are right will make it more specific in terms of ops conversion part.  Previous message may give misleading info.

File: ngraph/frontend/paddlepaddle/src/op/equal.cpp (hunk 3 above)
Comment (zhangYiIntel) [reply ngraph/frontend/paddlepaddle/src/op/equal.cpp:21]: Will implement equal like other elementwise  ops which considers the broadcast case

File: ngraph/frontend/paddlepaddle/src/op_table.cpp (hunk 1 above)
Comment (ilya-lavrenov) [reply ngraph/frontend/paddlepaddle/src/op_table.cpp:26]: We don't talk about custom operations here. There is a way how we can register our operations handlers without creation of multiple hpp files. No more

File: ngraph/frontend/paddlepaddle/src/op/flatten_contiguous_range.cpp (hunk 5 above)
Comment (zhangYiIntel) [reply ngraph/frontend/paddlepaddle/src/op/flatten_contiguous_range.cpp:50]: @liubo-intel  Please have check

File: ngraph/frontend/paddlepaddle/src/op_table.cpp (hunk 1 above)
Comment (nosovmik) [reply ngraph/frontend/paddlepaddle/src/op_table.cpp:26]: @zhangYiIntel We have short discussion, can you try the following approach (https://github.com/openvinotoolkit/openvino/pull/6632#discussion_r676377975)
1) Remove all op/*.hpp files
2) Add declarations only in op_table.cpp file as these are not needed anywhere else
3) In every op/*.cpp file - just include `#include <node_context.hpp>" as operation definition doesn't need to know it's forward declaration

Comment (zhangYiIntel) [reply ngraph/frontend/paddlepaddle/src/op_table.cpp:26]: @nosovmik No Problem, I will have a try.

File: ngraph/frontend/paddlepaddle/src/op/flatten_contiguous_range.cpp (hunk 5 above)
Comment (zhangYiIntel) [reply ngraph/frontend/paddlepaddle/src/op/flatten_contiguous_range.cpp:50]: @jane-intel I check the implementation, I think `special_zero = true` is we want.  We don't want a  zero dimensions. Here I paste the definition of Paddle `flatten_contiguous_range` for reference.

`A 2D tensor is reshaped input tensor. The input dimensions"
              "up to axis are flattened to the outer dimension of the output"
              "and the remaining input dimensions are flattened into the inner"
              "dimension of the output`

File: ngraph/frontend/paddlepaddle/src/op/expand_v2.cpp (hunk 4 above)
Comment (luo-cheng2021) [reply ngraph/frontend/paddlepaddle/src/op/expand_v2.cpp:17]: There is a conflict between [doc string](https://github.com/PaddlePaddle/Paddle/blob/758dd7bb8f560e45bc49855e4776a9d64c95aea9/paddle/fluid/operators/expand_v2_op.cc#L138) and the ['expand' document](https://www.paddlepaddle.org.cn/documentation/docs/en/api/paddle/expand_en.html) which says 'The dimension to expand must have a value 1.'
Anyway, I'll use the 'Tile' op to support all cases.

File: ngraph/frontend/paddlepaddle/src/op_table.cpp (hunk 1 above)
Comment (zhangYiIntel) [reply ngraph/frontend/paddlepaddle/src/op_table.cpp]: Fix in 5caa477f935e11eaab789c301af16866d6ad89b7

File: ngraph/frontend/paddlepaddle/src/op/dropout.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (mangguo321) [reply ngraph/frontend/paddlepaddle/src/op/dropout.cpp:25]: Done.

**Review by ilya-lavrenov — APPROVED** [review]

File: ngraph/frontend/paddlepaddle/src/op/argmax.cpp [hunk 10]
```diff
@@ -2,8 +2,8 @@
 // SPDX-License-Identifier: Apache-2.0
 //
 
-#include "argmax.hpp"
```
Comment (ilyachur) [inline ngraph/frontend/paddlepaddle/src/op/argmax.cpp:5]: Why did you remove `argmax.hpp` I didn't find any forward declarations? Maybe I miss something.

Comment (zhangYiIntel) [reply ngraph/frontend/paddlepaddle/src/op/argmax.cpp:5]: All forward declaration are now moved to #

https://github.com/zhangYiIntel/openvino/blob/81bbfeb0e80159b54af09b19d1d2afd100bca3b5/ngraph/frontend/paddlepaddle/src/op_table.cpp#L16

Comment (ilyachur) [reply ngraph/frontend/paddlepaddle/src/op/argmax.cpp:5]: ok, got it.

**Review by ilyachur — APPROVED** [review]

**Review by nosovmik — APPROVED** [review]

File: ngraph/frontend/paddlepaddle/src/op/transpose2.cpp [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+        {
+            namespace op
+            {
+                NamedOutputs transpose2(const NodeContext& node)
+                {
+                    auto data = node.get_ng_input("X");
+                    auto perm = node.get_attribute<std::vector<int>>("axis");
+
+                    auto rank =
+                        static_cast<unsigned long>(data.get_partial_shape().rank().get_length());
```
Comment (jane-intel) [inline ngraph/frontend/paddlepaddle/src/op/transpose2.cpp:22]: there is no check if rank is dynamic

**Review by jane-intel — CHANGES_REQUESTED** [review]

File: ngraph/frontend/paddlepaddle/src/op/greater_equal.cpp [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+    {
+        namespace pdpd
+        {
+            namespace op
+            {
+                NamedOutputs greater_equal(const NodeContext& node)
+                {
+                    auto x = node.get_ng_input("X");
+                    auto y = node.get_ng_input("Y");
+                    return node.default_single_output_mapping(
```
[the remark this answers is not in the document]
Comment (zhangYiIntel) [reply ngraph/frontend/paddlepaddle/src/op/greater_equal.cpp:21]: Yes, this one is already been implemented as eltwise. Forget to remove the cpp files. Will remove it.

File: ngraph/frontend/paddlepaddle/src/op/split.cpp [hunk 13]
```diff
@@ -24,7 +24,7 @@ namespace ngraph
                     {
                         auto input = node.get_ng_input("AxisTensor");
                         auto zero_node = Constant::create(element::i32, {1}, {0});
-                        axis = std::make_shared<Reshape>(input, zero_node, false);
+                        axis = std::make_shared<ReduceMin>(input, zero_node, false);
```
[the remark this answers is not in the document]
Comment (zhangYiIntel) [reply ngraph/frontend/paddlepaddle/src/op/split.cpp:27]: Yes Reshape throws error in this case.

File: ngraph/frontend/paddlepaddle/src/op/flatten_contiguous_range.cpp (hunk 5 above)
Comment (zhangYiIntel) [reply ngraph/frontend/paddlepaddle/src/op/flatten_contiguous_range.cpp:50]: Got it,  will set it to false

File: ngraph/frontend/paddlepaddle/src/op/transpose2.cpp [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+                        data.get_partial_shape().rank().is_static(),
+                        "[Frontend]Paddle transpose2 input data rank must be static");
+
+                    auto rank =
+                        static_cast<unsigned long>(data.get_partial_shape().rank().get_length());
+
+                    auto input_order =
+                        ngraph::opset6::Constant::create(ngraph::element::i64, {rank}, perm);
+                    return node.default_single_output_mapping(
+                        {std::make_shared<ngraph::opset6::Transpose>(data, input_order)}, {"Out"});
```
Comment (jane-intel) [inline ngraph/frontend/paddlepaddle/src/op/transpose2.cpp:32]:
```suggestion
                    auto data = node.get_ng_input("X");
                    auto perm = node.get_attribute<std::vector<int>>("axis");
                    auto input_order =
                        ngraph::opset6::Constant::create(ngraph::element::i64, {perm.size()}, perm);
                    return node.default_single_output_mapping(
                        {std::make_shared<ngraph::opset6::Transpose>(data, input_order)}, {"Out"});
```

Comment (zhangYiIntel) [reply ngraph/frontend/paddlepaddle/src/op/transpose2.cpp:32]: All right, but do we still need to check whether rank is static ?

Comment (jane-intel) [reply ngraph/frontend/paddlepaddle/src/op/transpose2.cpp]: No we don't

Comment (jane-intel) [reply ngraph/frontend/paddlepaddle/src/op/transpose2.cpp]: I suggested to delete it

Comment (zhangYiIntel) [reply ngraph/frontend/paddlepaddle/src/op/transpose2.cpp]: OK, will delete this useless check

**Review by jane-intel — APPROVED** [review]
